Migrate a part of the jobs API to Fast API

heisner-tillman opened the pull request:
This is a part of #10889.

## Summary
- [x] Refactored API routes
- [x] Added pydantic models
- [x] Operations now return pydanctic models
- [x] Removed the mapping to the legacy routes

## How to test the changes?
- [x] This is a refactoring of components with existing test coverage.
- [x] I added automated tests for the operations missing tests
- [x] Instructions for manual testing are as follows:
    You can find the interactive API documentation here: [http://127.0.0.1:8080/api/docs#/jobs](url)

![image](https://github.com/galaxyproject/galaxy/assets/117033448/da5dbd07-9b5b-4791-9b68-1dcf794a235b)



## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

heisner-tillman committed:
Remove redundant endpoints that have already been migrated

heisner-tillman committed:
Move method get_job to JobsService

heisner-tillman committed:
Refactor common_problems operation

heisner-tillman committed:
Remove mapping to the legacy route of common_problems operation

heisner-tillman committed:
Add pydantic model for common_problems operation

heisner-tillman committed:
Update client schema

heisner-tillman committed:
Create schema file for jobs API and move corresponding model to it

heisner-tillman committed:
Move dictify methods to JobsService

heisner-tillman committed:
Refactor resume operation

heisner-tillman committed:
Add pydantic model for dictify methods

heisner-tillman committed:
Remove mapping to the legacy route of resume operation

heisner-tillman committed:
Fix typing mistake

heisner-tillman committed:
Fix codestyle mistakes

heisner-tillman committed:
Update client schema

heisner-tillman committed:
Reorder endpoint operations

heisner-tillman committed:
Refactor error operation

heisner-tillman committed:
Add pydantic models for error operation

heisner-tillman committed:
Remove mapping to the legacy route of the error operation

heisner-tillman committed:
Update client schema

heisner-tillman committed:
Refactor error test

heisner-tillman committed:
Refactor search operation

heisner-tillman committed:
Add ServiceBase as base class to JobsService

heisner-tillman committed:
Remove mapping to legacy route

heisner-tillman committed:
Add part of the pydantic models for search operation

heisner-tillman committed:
Refactor search tests to specify payload as json

heisner-tillman committed:
Regenerate client schema

heisner-tillman committed:
Add descriptions to pydantic model

heisner-tillman committed:
Add TODO/suggestion to the search operation

heisner-tillman committed:
Refactor some JobModels to enable reuse

heisner-tillman committed:
Add pydantic models for search operation

heisner-tillman committed:
Reformulate TODO

heisner-tillman committed:
Regenerate client schema

heisner-tillman committed:
Refactor inputs operation

heisner-tillman committed:
Remove legacy route

heisner-tillman committed:
Add test for inputs operation

heisner-tillman committed:
Regenerate the client schema

heisner-tillman committed:
Refactor outputs operation

heisner-tillman committed:
Remove mapping to legacy route

heisner-tillman committed:
Refactor inputs test to include testing for outputs

heisner-tillman committed:
Regenrate client schema

heisner-tillman committed:
Refactor pydantic model for search operation

heisner-tillman committed:
Refactor search operation to return pydantic model

heisner-tillman committed:
Regenerate client schema

heisner-tillman committed:
Refactor delete operation

heisner-tillman committed:
Add pydantic model for delete operation

heisner-tillman committed:
Create internal method for Job testing and add test for delete operation

heisner-tillman committed:
Regenerate client schema

heisner-tillman committed:
Add proper documentation to show operation and use Annotated on delete body

heisner-tillman committed:
Refactor destination_params operation

heisner-tillman committed:
Add pydantic model for destination_params operation

heisner-tillman committed:
Add test for destination_params operation

heisner-tillman committed:
Remove mapping to the legacy route

heisner-tillman committed:
Regenerate the client schema

heisner-tillman committed:
Refactor part of the metrics operation

heisner-tillman committed:
Remove mapping to the legacy route

heisner-tillman committed:
Add test for part of the metrics operation

heisner-tillman committed:
Regenerate the client schema

heisner-tillman committed:
Refactor part of the parameters_display operation

heisner-tillman committed:
Remove mapping to legacy route

heisner-tillman committed:
Add tests for part of parameters_display operation

heisner-tillman committed:
Regenerate the client schema

heisner-tillman committed:
Refactor second part of the metrics operation

heisner-tillman committed:
Remove mapping to the legacy route

heisner-tillman committed:
Regenerate client schema

heisner-tillman committed:
Refactor get_job method

heisner-tillman committed:
Remove TODO from tests

heisner-tillman committed:
Refactor second part of parameters_display operation

heisner-tillman committed:
Remove mapping to legacy route

heisner-tillman committed:
Regenerate client schema

heisner-tillman committed:
Remove unused methods

heisner-tillman committed:
Remove unused dependencies of JobController

heisner-tillman committed:
Add TODO

heisner-tillman committed:
Add pydantic model in metrics operations return

heisner-tillman committed:
Remove outdated TODOs

heisner-tillman committed:
Regenerate the client schema

davelopez commented on lib/galaxy/schema/jobs.py:
```diff
@@ -0,0 +1,160 @@
+        description="Dictionary mapping all the tool outputs (by name) with the corresponding dataset information.",
+    )
+    # TODO add description, check type and add proper default
+    copied_from_job_id: Any = Field(default=None, title="Copied from Job-ID", description="?")
```
I guess this is just an optional encoded ID if the job is a copy from another.
```suggestion
    copied_from_job_id: Optional[EncodedDatabaseIdField] = Field(default=None, title="Copied from Job-ID", description="?")
```

heisner-tillman replied:
I made the suggested changes.

davelopez commented on lib/galaxy/schema/jobs.py:
```diff
@@ -0,0 +1,160 @@
+    )
+    # TODO add description, check type and add proper default
+    copied_from_job_id: Any = Field(default=None, title="Copied from Job-ID", description="?")
+    output_collections: Any = Field(default={}, title="Output collections", description="?")
```
Not sure of the type of this one, but maybe you can try to trace it from here:

https://github.com/galaxyproject/galaxy/blob/a8b32e9d095022b0238005d3506db09cd316166b/lib/galaxy/webapps/galaxy/services/tools.py#L222

mvdbeek replied:
I don't think this is the right source, it should be https://github.com/mvdbeek/galaxy/blob/4e15e6929ef241d4d6428f84be14955f98c29cc4/lib/galaxy/model/__init__.py#L1854-L1857. So a variant of `EncodedDatasetSourceId` just for `hdca` should be sufficient.

davelopez commented on lib/galaxy/schema/jobs.py:
```diff
@@ -0,0 +1,160 @@
+    # TODO add description, check type and add proper default
+    runner: Any = Field(default=Required, title="Runner", description="?", alias="Runner")
+    runner_job_id: Any = Field(default=Required, title="Runner Job ID", description="?", alias="Runner Job ID")
+    handler: Any = Field(default=Required, title="Handler", description="?", alias="Handler")
```
After grepping around a bit I think these 3 fields are strings

heisner-tillman replied:
Alright I typed them as string, but the question still remains what to add to the description. I will look into it and if I don't succeed ask again

davelopez commented on lib/galaxy/webapps/galaxy/api/jobs.py:
```diff
@@ -167,26 +181,27 @@
-        return self.service.show(trans, id, bool(full))
+    manager: JobManager = depends(JobManager)
+    hda_manager: HDAManager = depends(HDAManager)
+    job_search: JobSearch = depends(JobSearch)
```
You already have the `JobManager`, `HDAManager`, and `JobSearch` instances in your `JobService`.

In general, I recommend encapsulating the logic of the routes in the `JobService` so the API module only contains FastAPI documentation and basic parameter or response massaging. By keeping the logic in a function inside the service we can easily change in the future to a different API framework more cleanly.

heisner-tillman replied:
Thank you for reminding me :) I adjusted it

heisner-tillman committed:
Access managers via service, remove uneccessary comments/TODOs and add pydantic model to parameters_display_by_job_id return

heisner-tillman committed:
Add pydantic model for parameters_display_by_job_id operation

heisner-tillman committed:
Regenerate client schema

heisner-tillman committed:
Remove uneccessary comments and add pydantic model to prarameters_display_by_dataset return

heisner-tillman committed:
Extend JobDisplayParametersSummary to use it in parameters_display_by_dataset return

heisner-tillman committed:
Regenerate the client schema

heisner-tillman committed:
Regenerate the client schema

heisner-tillman commented on lib/galaxy/webapps/galaxy/api/jobs.py:
```diff
@@ -257,148 +266,262 @@ def common_problems(self, trans: ProvidesUserContext, id, **kwd):
-        job = self.__get_job(trans, **kwd)
-        return summarize_job_metrics(trans, job)
+        job = self.service.get_job(trans, job_id=id, hda_ldda=hda_ldda)
+        return summarize_job_parameters(trans, job)
```
The legacy endpoint had this large comment block attached to it. Is it still true or can I remove it?

heisner-tillman commented on lib/galaxy/webapps/galaxy/api/jobs.py:
```diff
@@ -257,148 +266,262 @@ def common_problems(self, trans: ProvidesUserContext, id, **kwd):
+            raise exceptions.ObjectNotFound("Requested tool not found")
+        # TODO the inputs are actually a dict, but are passed as a JSON dump
+        # maybe change it?
+        inputs = json.loads(payload.inputs)
```
The inputs are passed as a JSON string in the payload. Can I somehow decode the JSON string inside the pydantic model and then typecheck it before calling the actual endpoint logic?
This would produce a cleaner error or am I wrong?

mvdbeek replied:
Yes, a validator should allow you to do that (and also allow us to deprecate the unnecessary dump)

heisner-tillman replied:
Thank you for implementing this!

heisner-tillman commented on lib/galaxy/webapps/galaxy/api/jobs.py:
```diff
@@ -257,148 +266,262 @@ def common_problems(self, trans: ProvidesUserContext, id, **kwd):
+            message = payload.message
+        else:
+            message = None
+        return self.service.job_manager.stop(job, message=message)
```
This endpoint actually cancels a job, while this may lead to deletion, I think naming the endpoint delete is not really intuitive. I took the freedom to assign its name="cancel_job", but I do not want to rename the endpoint yet. Any suggestions?

mvdbeek replied:
Thanks, this is fine.

heisner-tillman replied:
So I can rename it?

mvdbeek replied:
No, i meant this is fine as is. We can discuss this outside of the PR context ... there's a mode in the equivalent old-style controller that also deletes job outputs.

heisner-tillman replied:
Ok thanks :)

heisner-tillman commented on lib/galaxy_test/api/test_jobs.py:
```diff
@@ -937,9 +937,97 @@ def test_job_build_for_rerun_list_list(self, history_id):
+            f"/api/datasets/{dataset_id}/parameters_display", data={"hda_ldda": "hda"}
+        )
+        self._assert_status_code_is(display_parameters_by_dataset_response, 200)
+        # TODO maybe extend test
```
So the endpoints "destination_params", "metrics" and "parameters_display" had no tests yet. I wrote tests for all of them, but they are really simple. Basically the only check that the endpoint works. It does not check the return data, as I do not really know anything about that.
Maybe they should be extended, any suggestions on if and how?

mvdbeek replied:
These are fairly internal, I think this is good enough for now!

heisner-tillman commented on lib/galaxy_test/api/test_jobs.py:
```diff
@@ -937,9 +937,97 @@ def test_job_build_for_rerun_list_list(self, history_id):
         )
         assert rerun_content == run_content
 
+    @pytest.mark.require_new_history
+    def test_get_inputs_and_outputs(self, history_id):
+        dataset_id = self.__history_with_ok_dataset(history_id)
+        inputs = json.dumps({"input1": {"src": "hda", "id": dataset_id}})
+        search_response = self._create_and_search_job(history_id, inputs, tool_id="cat1")
+        job_id = search_response.json()[0]["id"]
+        job_first_output_name, job_first_output_values = list(search_response.json()[0]["outputs"].items())[0]
+        # get the inputs of the job
+        job_response = self._get(f"jobs/{job_id}/inputs")
+        self._assert_status_code_is(job_response, 200)
+        job_first_input = job_response.json()[0]
+        # validate input response
+        assert job_first_input.get("name") == "input1"
+        assert job_first_input.get("dataset") == {"src": "hda", "id": dataset_id}
+        # get the outputs of the job
+        job_response = self._get(f"jobs/{job_id}/outputs")
+        self._assert_status_code_is(job_response, 200)
+        job_first_output = job_response.json()[0]
+        # validate output response
+        assert job_first_output.get("name") == job_first_output_name
+        assert job_first_output.get("dataset").get("id") == job_first_output_values.get("id")
+        assert job_first_output.get("dataset").get("src") == job_first_output_values.get("src")
```
The endpoints "inputs" and "outputs" also had no tests. So I wrote this. Any suggestions on making the tests more comprehensive or is it enough?

mvdbeek replied:
this is fine

mvdbeek commented on lib/galaxy/webapps/galaxy/api/jobs.py:
```diff
@@ -257,148 +266,262 @@ def common_problems(self, trans: ProvidesUserContext, id, **kwd):
-
-        :rtype:     list
-        :returns:   list containing job metrics
+        # TODO is this still true?
```
yes

mvdbeek commented on lib/galaxy/webapps/galaxy/api/jobs.py:
```diff
@@ -257,148 +266,262 @@ def common_problems(self, trans: ProvidesUserContext, id, **kwd):
-    @expose_api
-    def destination_params(self, trans: ProvidesUserContext, **kwd):
+    @router.get(
+        "/api/datasets/{id}/parameters_display",
```
```suggestion
        "/api/datasets/{dataset_id}/parameters_display",
```

mvdbeek commented on lib/galaxy/schema/jobs.py:
```diff
@@ -0,0 +1,268 @@
+    )
+
+
+class JobAssociation(Model):
```
These are the job input and output associations, can you split this into input and output models ?

heisner-tillman replied:
Yes

mvdbeek commented on lib/galaxy/schema/jobs.py:
```diff
@@ -0,0 +1,268 @@
+    name: str = Field(
+        default=Required,
+        title="name",
+        description="The name of the associated dataset.",
```
```suggestion
        description="Name of the job parameter.",
```
This is the input or output name as defined in the tool, the name of the dataset is something different

mvdbeek commented on lib/galaxy/schema/jobs.py:
```diff
@@ -0,0 +1,268 @@
+    dataset: EncodedDatasetSourceId = Field(
+        default=Required,
+        title="dataset",
+        description="The associated dataset.",
```
```suggestion
        description="Reference to the associated item.",
```

mvdbeek commented on lib/galaxy/schema/jobs.py:
```diff
@@ -0,0 +1,268 @@
+        title="name",
+        description="The name of the associated dataset.",
+    )
+    dataset: EncodedDatasetSourceId = Field(
```
EncodedDatasetSourceId.src is a union of `hda` and `ldda`, but `hdca`, `dce` and `dc` are also valid here. Can you add a new model `EncodedDataItemSourceId` and replace `DatasetSourceType` with `DataItemSourceType` ?

heisner-tillman replied:
DataItemSourceType was not defined anywhere, so I created it along with EncodedDataItemSourceId.

mvdbeek commented on lib/galaxy/schema/jobs.py:
```diff
@@ -0,0 +1,268 @@
+    dataset_id: DecodedDatabaseIdField = Field(
+        default=Required,
+        title="Dataset ID",
+        description="The dataset ID related to the error.",
```
```suggestion
        title="History Dataset Association ID",
        description="The History Dataset Association ID related to the error.",
```

eventually we should define aliases for the specific types of IDs, so we always know what type of id we're dealing with. But that's definitely a separate task!

mvdbeek commented on lib/galaxy/schema/jobs.py:
```diff
@@ -0,0 +1,268 @@
+    email: Optional[str] = Field(
+        default=None,
+        title="Email",
+        description="",
```
```suggestion
    email: Optional[str] = Field(
        default=None,
        title="Email",
        description="Email address for communication with the user. Only required for anonymous users.",
```

mvdbeek commented on lib/galaxy/schema/jobs.py:
```diff
@@ -0,0 +1,268 @@
+        default=Required,
+        title="Inputs",
+        description="The inputs of the job as a JSON dump.",
+    )
```
This is a super hard problem, there's not really a schema that we can generate statically as each tool allows different inputs. Can we just make this `Dict[str, Any]` ? If this fails we can attach a custom validator that does a json.loads on the payload.

mvdbeek commented on lib/galaxy/schema/jobs.py:
```diff
@@ -0,0 +1,268 @@
+        title="Inputs",
+        description="The inputs of the job as a JSON dump.",
+    )
+    state: str = Field(
```
Can you re-use this ? https://github.com/mvdbeek/galaxy/blob/4e15e6929ef241d4d6428f84be14955f98c29cc4/lib/galaxy/schema/schema.py#L87

heisner-tillman replied:
Yes.

mvdbeek commented on lib/galaxy/schema/jobs.py:
```diff
@@ -0,0 +1,268 @@
+    )
+
+    class Config:
+        extra = Extra.allow  # This is used for items named file_ and __file_
```
Can we allow these 2 then ?

mvdbeek replied:
Ah, these are just suffixes. We can revisit that another time, in principle we can do our own validation

mvdbeek commented on lib/galaxy/webapps/galaxy/api/jobs.py:
```diff
@@ -257,148 +266,262 @@ def common_problems(self, trans: ProvidesUserContext, id, **kwd):
+    )
+    def parameters_display_by_dataset(
+        self,
+        id: Annotated[DecodedDatabaseIdField, DatasetIdPathParam],
```
```suggestion
        dataset_id: Annotated[DecodedDatabaseIdField, DatasetIdPathParam],
```

mvdbeek commented on lib/galaxy/schema/jobs.py:
```diff
@@ -0,0 +1,268 @@
+
+
+class EncodedDatasetJobInfo(EncodedDatasetSourceId):
+    uuid: UUID4 = UuidField
```
Same comment applies, this can be more than just HistoryDatasetAssociation and LibraryDatasetDatasetAssociation, so `EncodedDatasetSourceId` is not correct, and `uuid` is optional (and would love if you can mark this as deprecated, it wasn't well-thought through when it was added to the API).

heisner-tillman replied:
So like this?
```

uuid: Optional[UUID4] = Field(
        default=None,
        deprecated=True,
        title="UUID",
        description="Universal unique identifier for this dataset. In this context the uuid is optional and marked as deprecated.",
    )
```

mvdbeek replied:
```python
uuid: Optional[UUID4] = Field(
        default=None,
        deprecated=True,
        title="UUID",
        description="Universal unique identifier for this dataset.",
    )
```
Optional and deprecated should be clear from the schema, don't need to repeat that.

mvdbeek commented on lib/galaxy/schema/jobs.py:
```diff
@@ -0,0 +1,268 @@
+    inputs: Dict[str, EncodedDatasetJobInfo] = Field(
+        {},
+        title="Inputs",
+        description="Dictionary mapping all the tool inputs (by name) with the corresponding dataset information.",
```
```suggestion
        description="Dictionary mapping all the tool inputs (by name) to the corresponding data references.",
```

mvdbeek commented on lib/galaxy/schema/jobs.py:
```diff
@@ -0,0 +1,268 @@
+    outputs: Dict[str, EncodedDatasetJobInfo] = Field(
+        {},
+        title="Outputs",
+        description="Dictionary mapping all the tool outputs (by name) with the corresponding dataset information.",
```
```suggestion
        description="Dictionary mapping all the tool outputs (by name) to the corresponding data references.",
```

mvdbeek commented on lib/galaxy/schema/jobs.py:
```diff
@@ -0,0 +1,268 @@
+    # TODO add description, check type and add proper default
+    copied_from_job_id: Optional[EncodedDatabaseIdField] = Field(
+        default=None, title="Copied from Job-ID", description="?"
+    )
```
```suggestion
    copied_from_job_id: Optional[EncodedDatabaseIdField] = Field(
        default=None, title="Copied from Job-ID", description="Reference to cached job if job execution was cached."
    )
```

mvdbeek commented on lib/galaxy/schema/jobs.py:
```diff
@@ -0,0 +1,268 @@
+
+class JobDestinationParams(Model):
+    # TODO add description, check type and add proper default
+    runner: str = Field(default=Required, title="Runner", description="?", alias="Runner")
```
```suggestion
    runner: str = Field(default=Required, title="Runner", description="Job runner class", alias="Runner")
```

mvdbeek commented on lib/galaxy/schema/jobs.py:
```diff
@@ -0,0 +1,268 @@
+
+
+class JobDestinationParams(Model):
+    # TODO add description, check type and add proper default
```
```suggestion
```

heisner-tillman replied:
The description for `handler` is still missing. Is handler also a class related to the job? If so can I just say:
`description="Job handler class"`?

mvdbeek replied:
You can also totally leave out the description if it's not adding anything beyond what you can guess from the attribute itself. In this case though it's `Name of the process that handled the job.`

mvdbeek commented on lib/galaxy/schema/jobs.py:
```diff
@@ -0,0 +1,268 @@
+class JobDestinationParams(Model):
+    # TODO add description, check type and add proper default
+    runner: str = Field(default=Required, title="Runner", description="?", alias="Runner")
+    runner_job_id: str = Field(default=Required, title="Runner Job ID", description="?", alias="Runner Job ID")
```
```suggestion
    runner_job_id: str = Field(default=Required, title="Runner Job ID", description="ID assigned to submitted job by external job running system", alias="Runner Job ID")
```

mvdbeek commented on lib/galaxy/schema/jobs.py:
```diff
@@ -0,0 +1,268 @@
+    output_collections: Any = Field(default={}, title="Output collections", description="?")
+
+
+class JobDestinationParams(Model):
```
I think these should all be nullable and that's a good default, the job won't have these if it's not queued yet.

mvdbeek commented on lib/galaxy/schema/jobs.py:
```diff
@@ -0,0 +1,268 @@
+
+class JobOutput(Model):
+    label: Any = Field(default=Required, title="Output label", description="The output label")  # check if this is true
+    value: EncodedDatasetSourceId = Field(default=Required, title="dataset", description="The associated dataset.")
```
This can also be more than just `ldda` and `hda`, see the comments above.

mvdbeek commented on lib/galaxy/schema/jobs.py:
```diff
@@ -0,0 +1,268 @@
+    )  # TODO check correctness
+    elements: List[JobTargetElement] = Field(
+        default=Required, title="Elements", description="List of job target elements"
+    )  # TODO check correctness
```
I think you're modelling the outputs of the `__FETCH_DATA__` upload tool here. I don't think it is correct to do that, since that won't work for anything but the upload tool

mvdbeek commented on lib/galaxy/schema/jobs.py:
```diff
@@ -0,0 +1,268 @@
+        description="The depth of the job parameter.",
+    )
+    value: Union[List[JobParameterValuesSimple], JobParameterValuesExtensive, str] = Field(
+        default=Required, title="Value", description="The values of the job parameter"
```
I don't think this can be modelled, it is different for ever tool you use. Make it Any ?

mvdbeek commented on lib/galaxy/schema/schema.py:
```diff
@@ -1820,19 +1820,22 @@ class JobIdResponse(Model):
 
 
-class JobBaseModel(Model):
+class JobIDs(Model):
```
This seems wrong, anything going out via the API has to be encoded. Why did you make that change ?

heisner-tillman replied:
So the original `JobBaseModel` contained, among others, the following attributes:
```
id: DecodedDatabaseIdField = EntityIdField
history_id: Optional[DecodedDatabaseIdField] = Field(
    None,
    title="History ID",
    description="The encoded ID of the history associated with this item.",
)
```
I moved them to the new model:
```
class JobIDs(Model):
    id: DecodedDatabaseIdField = EntityIdField
    history_id: Optional[DecodedDatabaseIdField] = Field(
        None,
        title="History ID",
        description="The encoded ID of the history associated with this item.",
    )
```
Now this model:
`Class JobSummary(JobBaseModel)`
does not contain :
```
id: DecodedDatabaseIdField = EntityIdField
history_id: Optional[DecodedDatabaseIdField] = Field(
    None,
    title="History ID",
    description="The encoded ID of the history associated with this item.",
)
```
This allows me to use `JobSummary` somewhere else, where I do not want the `decoded IDs`.

Every other model which originally inherits `JobSummary` or `JobBaseModel`:
- `class JobImportHistoryResponse(JobBaseModel, JobIDs)`
- `class JobDetails(JobSummary, JobIDs)`

Also inherits `JobIDs`, as the `decoded IDs` are not present in the new `JobSummary` or `JobBaseModel`

heisner-tillman replied:
Nonetheless you are right it is weird, that the original `JobBaseModel` contained `DecodedDatabaseIdField`s, as `JobImportHistoryResponse` is used as a response in `FastAPIHistories` (`lib/galaxy/webapps/galaxy/api/histories.py`).
`JobDetails` gets used all over the place and I do not know if it is ever returned.

Also the description and the actual data type in the attribute `history_id` of the original `JobBaseModel` do not match!

This is not my doing as I did not create `JobBaseModel`, `JobImportHistoryResponse` or `JobDetails` .

mvdbeek replied:
So can you undo the splitting and fix the ids to be `EncodedDatabaseIdField` ? I think this was correct previously minus the wrong field type

mvdbeek commented on lib/galaxy/webapps/galaxy/api/jobs.py:
```diff
@@ -257,148 +266,262 @@ def common_problems(self, trans: ProvidesUserContext, id, **kwd):
+    def parameters_display_by_job(
+        self,
+        id: Annotated[DecodedDatabaseIdField, JobIdPathParam],
+        hda_ldda: Annotated[DatasetSourceType, HdaLddaQueryParam] = DatasetSourceType.hda,
```
This parameter doesn't make sense here, we already require a job id. can you make this optional and mark it as deprecated?

heisner-tillman replied:
Yes

mvdbeek commented on lib/galaxy/webapps/galaxy/api/jobs.py:
```diff
@@ -257,148 +266,262 @@ def common_problems(self, trans: ProvidesUserContext, id, **kwd):
+        return self.service.dictify_associations(trans, job.output_datasets, job.output_library_datasets)
+
+    @router.get(
+        "/api/jobs/{id}/parameters_display",
```
```suggestion
        "/api/jobs/{job_id}/parameters_display",
```

mvdbeek commented on lib/galaxy/webapps/galaxy/api/jobs.py:
```diff
@@ -257,148 +266,262 @@ def common_problems(self, trans: ProvidesUserContext, id, **kwd):
+    )
+    def parameters_display_by_job(
+        self,
+        id: Annotated[DecodedDatabaseIdField, JobIdPathParam],
```
```suggestion
        job_id: Annotated[DecodedDatabaseIdField, JobIdPathParam],
```

mvdbeek commented on lib/galaxy/webapps/galaxy/api/jobs.py:
```diff
@@ -257,148 +266,262 @@ def common_problems(self, trans: ProvidesUserContext, id, **kwd):
-    @expose_api_anonymous
-    def parameters_display(self, trans: ProvidesUserContext, **kwd):
+    @router.get(
+        "/api/jobs/{id}/metrics",
```
```suggestion
        "/api/jobs/{job_id}/metrics",
```

mvdbeek commented on lib/galaxy/webapps/galaxy/api/jobs.py:
```diff
@@ -257,148 +266,262 @@ def common_problems(self, trans: ProvidesUserContext, id, **kwd):
+    )
+    def metrics_by_job(
+        self,
+        id: Annotated[DecodedDatabaseIdField, JobIdPathParam],
```
```suggestion
        job_id: Annotated[DecodedDatabaseIdField, JobIdPathParam],
```

mvdbeek commented on lib/galaxy/webapps/galaxy/api/jobs.py:
```diff
@@ -257,148 +266,262 @@ def common_problems(self, trans: ProvidesUserContext, id, **kwd):
+    def metrics_by_job(
+        self,
+        id: Annotated[DecodedDatabaseIdField, JobIdPathParam],
+        hda_ldda: Annotated[DatasetSourceType, HdaLddaQueryParam] = DatasetSourceType.hda,
```
```suggestion
        hda_ldda: Annotated[DatasetSourceType, HdaLddaQueryParam] = DatasetSourceType.hda,
```
we're not using this, can you make it optional and deprecate it ?

heisner-tillman replied:
Yes

mvdbeek commented on lib/galaxy/webapps/galaxy/api/jobs.py:
```diff
@@ -257,148 +266,262 @@ def common_problems(self, trans: ProvidesUserContext, id, **kwd):
+        return [JobMetric(**metric) for metric in summarize_job_metrics(trans, job)]
+
+    @router.get(
+        "/api/datasets/{id}/metrics",
```
```suggestion
        "/api/datasets/{dataset_id}/metrics",
```

mvdbeek commented on lib/galaxy/webapps/galaxy/api/jobs.py:
```diff
@@ -257,148 +266,262 @@ def common_problems(self, trans: ProvidesUserContext, id, **kwd):
+    )
+    def metrics_by_dataset(
+        self,
+        id: Annotated[DecodedDatabaseIdField, DatasetIdPathParam],
```
```suggestion
        dataset_id: Annotated[DecodedDatabaseIdField, DatasetIdPathParam],
```

mvdbeek commented on lib/galaxy/webapps/galaxy/api/jobs.py:
```diff
@@ -257,148 +266,262 @@ def common_problems(self, trans: ProvidesUserContext, id, **kwd):
+        name="get_metrics",
+        summary="Return job metrics for specified job.",
+    )
+    def metrics_by_dataset(
```
I think it makes sense to mark all the `*_by_dataset` as deprecated routes, they don't work for dataset collections and it's relatively indirect compared to getting information via the job.

mvdbeek commented on lib/galaxy/webapps/galaxy/api/jobs.py:
```diff
@@ -257,148 +266,262 @@ def common_problems(self, trans: ProvidesUserContext, id, **kwd):
-        :type   job_id: string
-        :param  job_id: Encoded job id
+    @router.get(
+        "/api/jobs/{id}",
```
```suggestion
        "/api/jobs/{job_id}",
```

mvdbeek commented on lib/galaxy/webapps/galaxy/api/jobs.py:
```diff
@@ -257,148 +266,262 @@ def common_problems(self, trans: ProvidesUserContext, id, **kwd):
+    )
+    def show(
+        self,
+        id: Annotated[DecodedDatabaseIdField, JobIdPathParam],
```
```suggestion
        job_id: Annotated[DecodedDatabaseIdField, JobIdPathParam],
```

mvdbeek commented on lib/galaxy/webapps/galaxy/api/jobs.py:
```diff
@@ -257,148 +266,262 @@ def common_problems(self, trans: ProvidesUserContext, id, **kwd):
+    # TODO find the mapping of the legacy route
+    # TODO rename? --> function cancels/stops job (no deletion involved?)
+    @router.delete(
+        "/api/jobs/{id}",
```
```suggestion
        "/api/jobs/{job_id}",
```

mvdbeek commented on lib/galaxy/webapps/galaxy/api/jobs.py:
```diff
@@ -257,148 +266,262 @@ def common_problems(self, trans: ProvidesUserContext, id, **kwd):
-        :type   dataset_id: string
-        :param  dataset_id: Encoded HDA or LDDA id
+    # TODO find the mapping of the legacy route
+    # TODO rename? --> function cancels/stops job (no deletion involved?)
```
```suggestion
```

methods that match the CRUD keywords are added automatically via https://github.com/mvdbeek/galaxy/blob/4e15e6929ef241d4d6428f84be14955f98c29cc4/lib/galaxy/webapps/galaxy/buildapp.py#L987 ... you should be able to delete this

heisner-tillman committed:
Apply some suggestions

heisner-tillman committed:
Apply some suggestions on pydantic models

heisner-tillman committed:
Refined EncodedJobDetails model

heisner-tillman committed:
Add description to JobDestinationParams model and remove TODOs

heisner-tillman committed:
Regenerate client schema

heisner-tillman committed:
Mark *_by_dataset operations as deprecated

heisner-tillman committed:
Mark some HdaLddaQueryParam's as deprecated

heisner-tillman committed:
Follow up on previous deprecation and refactor create operation

heisner-tillman committed:
Remove mapping to some legacy routes

heisner-tillman committed:
Regenerate client schema

heisner-tillman committed:
Divide JobAssociation into Input and Output model

heisner-tillman committed:
Regenerate the client schema

mvdbeek committed:
Use validator to load job search inputs

mvdbeek commented on lib/galaxy_test/api/test_jobs.py:
```diff
@@ -1008,3 +1093,4 @@ def __jobs_index(self, **kwds):
         jobs = jobs_response.json()
         assert isinstance(jobs, list)
         return jobs
+        return jobs
```
```suggestion
```

heisner-tillman committed:
Remove accidental code insertion in test_jobs.py

heisner-tillman committed:
Set Ids of JobBaseModel to encoded and undo previous model split

heisner-tillman committed:
Refine EncodedHdacaSourceId model

heisner-tillman committed:
Regenerate client schema

heisner-tillman committed:
Fix codestyle issue

heisner-tillman committed:
Remove completed TODOs and clarify some path parameters

heisner-tillman committed:
Remove outdated TODO

heisner-tillman committed:
Add test for delete operation

heisner-tillman committed:
Regenerate the client schema

heisner-tillman commented on lib/galaxy/webapps/galaxy/services/jobs.py:
```diff
@@ -92,3 +101,37 @@ def _check_nonadmin_access(
+            else:
+                dataset_instance = self.hda_manager.ldda_manager.get(trans, id=dataset_id)
+        # TODO Raise error if no ID passed? Never happens when called from Job API endpoints
+        return dataset_instance.creating_job
```
Should I raise an error here? I mean in theory this method is callable without passing any ID.

mvdbeek replied:
Yeah, sounds good.

heisner-tillman commented on lib/galaxy_test/api/test_jobs.py:
```diff
@@ -937,9 +937,95 @@ def test_job_build_for_rerun_list_list(self, history_id):
+        assert job_first_output.get("dataset").get("id") == job_first_output_values.get("id")
+        assert job_first_output.get("dataset").get("src") == job_first_output_values.get("src")
+
+    @pytest.mark.require_new_history
+    def test_delete_job(self, history_id):
+        dataset_id = self.__history_with_ok_dataset(history_id)
+        inputs = json.dumps({"input1": {"src": "hda", "id": dataset_id}})
+        search_payload = self._search_payload(history_id=history_id, tool_id="cat1", inputs=inputs)
+        # create a job
+        tool_response = self._post("tools", data=search_payload)
+        job_id = tool_response.json()["jobs"][0]["id"]
+        # delete the job without message
+        delete_job_response = self._delete(f"jobs/{job_id}")
+        self._assert_status_code_is(delete_job_response, 200)
+        assert delete_job_response.json() is True
+        # now that we deleted the job we should not find it anymore
+        search_payload = self._search_payload(history_id=history_id, tool_id="cat1", inputs=inputs)
+        empty_search_response = self._post("jobs/search", data=search_payload, json=True)
+        self._assert_status_code_is(empty_search_response, 200)
+        assert len(empty_search_response.json()) == 0
```
This test is for the delete operation. Is it extensive enough?

mvdbeek replied:
Yes, that's fine, thank you!

mvdbeek commented on lib/galaxy/schema/jobs.py:
```diff
@@ -0,0 +1,233 @@
+    )
+    handler: str = Field(
+        default=Required, title="Handler", description="Name of the process that handled the job.", alias="Handler"
+    )
```
```suggestion
    runner: Optional[str] = Field(None, title="Runner", description="Job runner class", alias="Runner")
    runner_job_id: Optional[str] = Field(
        None,
        title="Runner Job ID",
        description="ID assigned to submitted job by external job running system",
        alias="Runner Job ID",
    )
    handler: Optional[str] = Field(
        None, title="Handler", description="Name of the process that handled the job.", alias="Handler"
    )
```

heisner-tillman committed:
Add expetion in case of missing parameter

heisner-tillman committed:
Refine the JobDestinationParams model

heisner-tillman committed:
Regenerate the client schema

mvdbeek committed:
Drop POST /api/jobs

This is a remnant from the old routes setup that would automatically
create thr CRUD methods ... much better to just drop this.

mvdbeek committed:
Update description of '/api/jobs/{job_id}/parameters_display'

mvdbeek committed:
Drop BaseModel docstring

mvdbeek reviewed (approved):
Very nice work, thanks so much @heisner-tillman!

nsoranzo commented on lib/galaxy/schema/jobs.py:
```diff
@@ -0,0 +1,233 @@
+        title="Inputs",
+        description="The inputs of the job.",
+    )
+    state: JobState = Field(
+        default=Required,
```
This was previously optional, and now breaks a BioBlend test: https://github.com/nsoranzo/bioblend/actions/runs/6970136619/job/18968153613#step:10:3258

Can it be made optional again?
